chia/full_node/block_store.py

```diff
@@ -353,8 +357,7 @@ async def get_random_not_compactified(self, number: int) -> List[int]:
         # certain height are not compact. And if we do have compact orphan blocks, then all that
         # happens is that the occasional chain block stays uncompact - not ideal, but harmless.
         cursor = await self.db.execute(
-            f"SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 "
-            f"ORDER BY RANDOM() LIMIT {number}"
+            f"SELECT height FROM full_blocks WHERE is_fully_compactified=0 ORDER BY RANDOM() LIMIT {number}"
```

Comment:
To be quite honest, I don't really understand this change. your new query looks simple and straight forward. I don't know why the existing code groups by `height` not why it's computing the sum of `is_fully_compactified`.

The original query was introduced [here](https://github.com/Chia-Network/chia-blockchain/commit/a59073fc38196134b66848c617d23df59b0db33a). To address [this](https://github.com/Chia-Network/chia-blockchain/pull/7175) issue.

perhaps @beechord has some comments

Reply:
The aim of this query, from what I understood is, that you just 
select a random `{number}` of heights for not yet compacitified rows (=`is_full_compactified=0`).

The main benefit of this change is, that the index on is_fully_compactified can be used instead of using index on height.

Reply:
The basic problem here is, because consensus requires you to deal with forks and orphan blocks, blocks in the DB are not uniquely identified by height.

In my original PR, back when blueboxes still scanned work from the bottom up, I had to avoid naively compacting the oldest uncompact block because bluebox work would be wasted in multiple ways when compacting orphan blocks, first because `respond_compact_vdf` (iirc) would refuse to insert the compacted vdf for the orphan involved, second because even if you modified it, the network wouldn't gossip the proof because it belongs to an orphan that most people don't have. As this query is mostly run on bluebox timelords once every ~five minutes in the recommended settings, it's okay to spend a couple more seconds every five minutes if it means that you don't get stuck in an arbitrary part of the chain compacting proofs that nobody can use, at several minutes of CPU time per proof. And this was not just a theoretical problem but actually cropped up in practice by then. The more complicated query with the `HAVING` clause sidestepped all this.

A bit later in #8280, the query was modified again to return not the lowest uncompact block that was not definitely an orphan, but n random blocks that are not definitely orphans. I am not sure whether an index would even particularly help with the new query, as I believe `ORDER BY RANDOM()` computes a random number for every matching row to then choose those with the lowest, so again it is a scan through currently about 2/3 of the chain (https://www.arewecompactifiedyet.live/), index or not. Maybe a per-node random seed would help to make the ordering reproducible and therefore indexable, but that only really becomes necessary for a catchup protocol as detailed in #7678 where regular resource-constrained nodes run this query.
If you go and remove the `HAVING` here now, I believe but can't guarantee that blueboxing will basically keep working for now, but there will be more and more errors in the log as the network nears full compaction.

tl;dr this is not an easy fix, and it helps only a subset of nodes which is small and tends to consist of plotter-tier machines.

I've been busy with my dayjob recently so I don't know if there are even more current changes, but @fchirica knows the same things I do here and should be more up-to-speed than me.

Reply:
Very well explained. I didn't see the heights not beeing unique.

An improvement should still be possible by forcing the query to use index is_fully_compactified.
Example from my plotter:
```
sqlite> SELECT height FROM full_blocks indexed by is_fully_compactified GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30;
QUERY PLAN
|--SCAN TABLE full_blocks USING INDEX is_fully_compactified
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
Run Time: real 66.405 user 1.707902 sys 11.759515
...
sqlite> SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30;
QUERY PLAN
|--SCAN TABLE full_blocks USING INDEX full_block_height
`--USE TEMP B-TREE FOR ORDER BY
Run Time: real 140.793 user 2.701341 sys 18.591463
...
```

But as you mentioned, only high-tier nodes are effected.

Reply:
Hmm. Reading through the rest of the PR, it surprises me how fast `ORDER BY RANDOM()` is, so the currently unavoidable part is not as slow as I thought. If you can figure out a fast way of tagging orphan blocks for purposes of blueboxing, not consensus (so with relaxed requirements), you can index on is_orphan and height and hopefully get something reasonable.
The amount the right column is over 100k is the amount of orphans you have to deal with (the longer you stay online, the more you have):
```
sqlite> select (height - height%100000) as modheight,sum(is_fully_compactified),round(avg(length(block))),count() from full_blocks group by modheight;
0|100000|3425.0|100000
100000|99669|3426.0|100015
200000|99905|4474.0|101080
300000|49662|5056.0|100376
400000|13870|5219.0|100137
500000|12352|11862.0|100133
600000|12516|11793.0|100046
700000|12302|13666.0|100062
800000|9410|14518.0|100038
900000|8474|13404.0|100042
1000000|5640|17661.0|100247
1100000|1101|13196.0|63894
```

Reply:
I'm certainly missing some important information. 
How do you identify orphaned blocks using the current query?

I tried to reproduce on my side, but the only connection I see would be by matching prev_hash column from block_records table with header_hash column on full_blocks table.

Reply:
I don't - I merely know that when there's more than one block for a given height, all but one are orphans. I'm not aware of a way of identifying (as opposed to counting) orphans purely through a query, but I don't think I've ever tried to pull a table other than full_blocks into this. The way the rest of the implementation does it is for example by calling Blockchain.height_to_hash in FullNode._replace_proof: https://github.com/Chia-Network/chia-blockchain/blob/0d84578999c17c1ecd12dcc1bf4e0312a2f0b567/chia/full_node/full_node.py#L2078-L2082

If a consensus-strength is_orphan field existed, consensus code would need to make sure it was accurate and up-to-date every time a chain overtook another, which I presume is the reason it doesn't exist.

Reply:
> If a consensus-strength is_orphan field existed, consensus code would need to make sure it was accurate and up-to-date every time a chain overtook another, which I presume is the reason it doesn't exist.

I believe it was a mistake to not include a field like this in the original schema. I haven't implemented one yet, but I intend to experiment with this (along with a bunch of other schema improvements) for a v2 database format. I was envisioning a field `in_main_chain` or something like that (i.e. the inverse of `is_orphaned`). Keeping it up to date should not be any more complex than the height-to-hash map we currently keep up to date.

Reply:
I have two suggestions how to proceed with the query in this PR.

1. Change it to just use the better index (is_fully_compactified)
```
SELECT height FROM full_blocks 
INDEXED BY is_fully_compactified 
GROUP BY height 
HAVING sum(is_fully_compactified)=0 
ORDER BY random() LIMIT 30;
```
2. As long as there is no column like `in_main_chain` available and if the main chain defines itself backwards by `prev_hash` from `max(height) from block_records`, then we could use a recursive CTE to get the uncompactified heights only of the main chain 
```
WITH RECURSIVE
	fb_blockchain(header_hash, height) AS
	(
		SELECT prev_hash, height-1 
                   FROM block_records 
                   WHERE height = (SELECT MAX(height) FROM block_records)
		UNION ALL
		SELECT block_records.prev_hash, block_records.height 
                   FROM block_records, fb_blockchain 
		   WHERE fb_blockchain.header_hash = block_records.header_hash 
		   ORDER BY block_records.height DESC
	)
SELECT fb.height 
   FROM fb_blockchain bc, full_blocks fb 
   WHERE bc.header_hash = fb.header_hash AND fb.is_fully_compactified=0 
   ORDER BY random() LIMIT 30;
```
Execution time on my plotter(with empty fs cache):
`Run Time: real 137.201 user 4.742898 sys 4.824835`

Execution time (without clearing fs cache):
`Run Time: real 5.522 user 3.670922 sys 1.828944`

Reply:
A cheap way to check whether a block is in the chain or not is to use the height to hash map

Reply:
With my lack of python knowledge and time constraint, I don't see me figuring out how to use the height_to_hash map for this any time soon.

But I think the following SQL should do the work as well:
_(Fixed wrong height handling and used sligthly faster initial query)_
```
WITH RECURSIVE
	blockchain(header_hash, height) AS
	(
		SELECT prev_hash, height-1 
                   FROM block_records 
                   WHERE is_peak = 1
		UNION ALL
		SELECT block_records.prev_hash, blockchain.height-1 
                   FROM block_records, blockchain
		   WHERE blockchain.header_hash = block_records.header_hash
		   ORDER BY blockchain.height-1 DESC
	)
SELECT fb.height FROM blockchain bc, full_blocks fb 
	WHERE bc.header_hash = fb.header_hash AND fb.is_fully_compactified = 0 
	ORDER BY random() LIMIT 30;
```

Reply:
does this look right? https://github.com/Chia-Network/chia-blockchain/pull/9351
I attributed the commit to you

Reply:
Since I don't expect anybody to run a bluebox on a spinner, forcing index use looks fine to me. No objections from me. Eagerly awaiting an in_main_chain field, as I expect this to greatly simplify a compact-catchup implementation.